File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.java
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return buildBST(nums, 0, nums.length - 1);
    }
    
    private TreeNode buildBST(int[] nums, int left, int right) {
        // Base case: if left > right, return null (empty subtree)
        if (left > right) {
            return null;
        }
        
        // Find middle element to make it the root
        int mid = left + (right - left) / 2;
        
        // Create root node with the middle element
        TreeNode root = new TreeNode(nums[mid]);
        
        // Recursively build left and right subtrees
        root.left = buildBST(nums, left, mid - 1);
        root.right = buildBST(nums, mid + 1, right);
        
        return root;
    }
}

File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.rs
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        Self::build_bst(&nums, 0, nums.len() as i32 - 1)
    }
    
    fn build_bst(nums: &Vec<i32>, left: i32, right: i32) -> Option<Rc<RefCell<TreeNode>>> {
        // Base case: if left > right, return None (empty subtree)
        if left > right {
            return None;
        }
        
        // Find middle element to make it the root
        let mid = left + (right - left) / 2;
        
        // Create root node with the middle element
        let mut root = TreeNode::new(nums[mid as usize]);
        
        // Recursively build left and right subtrees
        root.left = Self::build_bst(nums, left, mid - 1);
        root.right = Self::build_bst(nums, mid + 1, right);
        
        // Return the tree wrapped in Option<Rc<RefCell<>>>
        Some(Rc::new(RefCell::new(root)))
    }
}